cpp/src/arrow/compute/row/compare_internal_avx2.cc

```diff
@@ -251,6 +253,35 @@ uint32_t KeyCompare::CompareBinaryColumnToRowHelper_avx2(
   }
 }
 
+namespace {
+
+/// Intrinsics `_mm256_i32gather_epi32/64` treat the `vindex` as signed integer, and we
+/// are using `uint32_t` to represent the offset, in range of [0, 4G), within the row
+/// table. When the offset is larger than `0x80000000` (2GB), those intrinsics will treat
+/// it as negative offset and gather the data from undesired address. To avoid this issue,
+/// we normalize the addresses by translating `base` `0x80000000` higher, and `offset`
+/// `0x80000000` lower. This way, the offset is always in range of [-2G, 2G) and those
+/// intrinsics are safe.
+
+constexpr auto two_gb = 0x80000000ull;
+
+template <int scale>
```

Comment:
Two things:
1) if we're using unsigned arithmetic below, the scale type should probably be unsigned for readability and sanity?
2) naming convention: can we make this `kScale`?

Reply:
1. The type of the third formal parameter of `_mm256_set1_epi32/64` is `int` so I'm just using `int` too.
2. Yeah, will do.

Reply:
Addressed.